Feature/hab 19 slides

VikkiRy opened the pull request:
Requirements:
1. Если пользователь с последнего слайда возвращается на предыдущий слайд, то активный элемент запуска регистрации не перестает быть активным.
2. Пользователь перелистывает слайды с помощью скролла по экрану.

VikkiRy committed:
- let nextScreen name and type were changed
- var signUpScreen was changed

VikkiRy committed:
en developmentRegion was removed

VikkiRy committed:
Merge branch 'develop' into feature/HAB-19-slides

# Conflicts:
#	BadHabbit/BadHabbit/UserStories/DI/StoriesAssembly.swift
#	BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpScreenRouter.swift

VikkiRy committed:
- new color assets were added
- swipe gestures and page control were added
- startButton was updated
- logoLabel was removed
- SignUpScreenRouter and SignUpScreenAssembly were updated

VikkiRy committed:
@objc funk parameters name were changed

VikkiRy committed:
сurrentPageIndicatorImage changed

VikkiRy committed:
logoImageViewSwiped parameter was renamed

VikkiRy committed:
Merge branch 'develop' into feature/HAB-19-slides

# Conflicts:
#	BadHabbit/BadHabbit.xcodeproj/project.pbxproj
#	BadHabbit/BadHabbit/SupportFiles/Assets.xcassets/signUpScreenLogo.imageset/Contents.json
#	BadHabbit/BadHabbit/SupportFiles/Assets.xcassets/signUpScreenLogo.imageset/logo.pdf

VikkiRy committed:
mountains.png was changed to mountains.pdf

nikGayko commented:
Could you record a video of swiping and attach to pr?

VikkiRy committed:
Merge branch 'develop' into feature/HAB-19-slides

# Conflicts:
#	BadHabbit/BadHabbit/SupportFiles/Assets.xcassets/Images/mountains.imageset/mountains.pdf
#	BadHabbit/BadHabbit/SupportFiles/Assets.xcassets/Images/mountains.imageset/mountains.png
#	BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift

VikkiRy committed:
- color assets were added
- SignUpScreenPresenter and SignUpViewController were updated
- xcassets enum name was updated in swiftgen.yml

VikkiRy commented:
https://github.com/The-Lessoner/bad_habbit/assets/97001077/6437a1a9-2c7e-443c-836f-9a2773570d76

VikkiRy commented:
> Could you record a video of swiping and attach to pr?

https://github.com/The-Lessoner/bad_habbit/assets/97001077/488d6698-28b8-4cda-9407-807c97db681a

nikGayko commented:
Судя по видосу никакого свайпа нет, только меняется позиция индикатора, какой элемент был такой и остался

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/DesignKitAssembly.swift:
```diff
@@ -30,5 +30,4 @@ final class DesignKitAssembly {
     var dropDownViewPresenter: DropDownViewPresenterAssemblyProtocol {
         DropDownViewPresenterAssembly()
     }
-
```
Try to avoid changes empty lines changes in files where you haven't any meaningful changes. That could cause extra conflicts and increment changes files count in PR(people don't like to review PR's with a lot of changes)

VikkiRy replied:
Got it. I removed this fix.

VikkiRy commented:
> Судя по видосу никакого свайпа нет, только меняется позиция индикатора, какой элемент был такой и остался

Просто еще не готовы слайды у дизайнера чтобы их применить... Мне добавить анимацию какую-нибудь чтобы было видно движение (при этом изображение будет 1 и то же)?

VikkiRy committed:
- backgroundGradientLayer and colors sets were added
- addPageControl() was refactored

VikkiRy committed:
SignUpViewController view was updated

VikkiRy committed:
startButton was updated

VikkiRy committed:
Created slidesCollectoinView instead of logoImageView

VikkiRy committed:
slidesCollectoinView constraints were changed

VikkiRy commented:
> Судя по видосу никакого свайпа нет, только меняется позиция индикатора, какой элемент был такой и остался


https://github.com/The-Lessoner/bad_habbit/assets/97001077/56681ad3-2c5a-4e24-b5bb-2fe5ee751745

VikkiRy committed:
SwipeGestureRecognizer.swift was removed

VikkiRy committed:
slidesCollectoinView centerY constraint was updated

HKurylchykava commented on BadHabbit/BadHabbit/SupportFiles/Colors.xcassets/pirpleLightColor.colorset/Contents.json:
rename colorset, please, purpleLightColor

VikkiRy committed:
Unused assets were removed

VikkiRy committed:
minimumLineSpacing value was changed

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -103,15 +113,34 @@ final class SignUpViewController: BaseViewController, SignUpViewProtocol {
-        buttonConfiguration.baseBackgroundColor = Asset.Colors.startButtonBaseBackground.color
-
-        startButton.configuration = buttonConfiguration
+        startButton.layer.cornerRadius = 12
```
у нас должно быть это значение в константах

VikkiRy replied:
Done

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -133,14 +162,16 @@ final class SignUpViewController: BaseViewController, SignUpViewProtocol {
-            make.centerX.centerY.equalTo(logoImageView)
+        pageControl.snp.makeConstraints { make in
+            make.centerX.equalTo(safeArea)
+            make.bottom.equalTo(phraseLabel.snp.top).inset(-30)
```
.offset(30)?

VikkiRy replied:
.offset(30) does the same in this case. Constraint was changed to equalTo(phraseLabel.snp.bottom).inset(93)

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/BackgroundGradientView.swift:
```diff
@@ -0,0 +1,27 @@
+
+import UIKit
+
+final class BackgroundGradientView: GradientView {
```
А как ты планируешь переиспользовать `GradientView` в будущем?

VikkiRy replied:
Посмотрела дизайн - пока больше нигде не понадобится переиспользование. Удалила этот класс

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/Models/LayoutConstants.swift:
```diff
@@ -15,7 +15,10 @@ enum LayoutConstants {
     enum ImageLogo {
         static let size: CGFloat = 196
+        static let height: CGFloat = 196
+        static let widht: CGFloat = 196
```
есть же size

VikkiRy replied:
Removed

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpScreenCollectionViewCell.swift:
```diff
@@ -0,0 +1,30 @@
+
+class SignUpScreenCollectionViewCell: UICollectionViewCell {
+    static let identifier = "SignUpScreenCollectionViewCell"
+    lazy var imageView = UIImageView()
```
не забывай модификаторы доступа

VikkiRy replied:
class was marked as final

nikGayko replied:
это не модификатор доступа) Я хотел сказать чтобы ты сделала imageView private

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpScreenPresenter.swift:
```diff
@@ -5,22 +5,63 @@
+    lazy var images: [UIImage] = {
+        var image: [UIImage] = []
+        (1...3).forEach { _ in
+            image.append(Assets.Images.signUpScreenLogo.image)
```
эти три элемента - это затычка пока нет настоящих изображений?

VikkiRy replied:
Верно

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpScreenPresenter.swift:
```diff
@@ -5,22 +5,63 @@
+    }()
+    
     private let router: SignUpRouterProtocol
+    private var startButtonIsEnabled = false {
```
на мой взгляд переменная лишняя - по сути ты дублируешь состояние одного элемента дважды, у тебя есть у UIButton переменная isEnabled и здесь еще одна. обе отражают одно и тоже состояние одного и тоже элемента, лучше этого избежать. Хорошо было бы здесь увидеть get set которые берут значение в этой кнопке, но тут ее нет, поэтому давай удалим

VikkiRy replied:
Свойство удалено

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpScreenPresenter.swift:
```diff
@@ -5,22 +5,63 @@
         router.presentSignInAppleScreen()
     }
+    
+    private func isNeedStartButtonUpdate(forPage page: Int) {
```
плохое название функции

VikkiRy replied:
Функция удалена

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpScreenPresenter.swift:
```diff
@@ -5,22 +5,63 @@
     }
+    
+    private func isNeedStartButtonUpdate(forPage page: Int) {
+        if page == 2 && startButtonIsEnabled == false {
```
я бы написал view.updateStartButton(isEnabled: page == 2

VikkiRy replied:
🔥  Круто, не помню, чтобы где-то встречала такое решение

VikkiRy replied:
func isNeedStartButtonUpdate удалена

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -8,17 +8,21 @@
+protocol SignUpViewProtocol: AnyObject {
+    func updateStartButton(isEnabled: Bool)
+    func updatePageControlPage(toPage page: Int)
+    func updateCollectionView(toPage page: Int)
```
добавляю возможность настраивать и controlPage и collectionView извне ты даешь возможность сломать состояние и сделать так чтобы в коллекции был 0-вой элемент, а на индикаторе 2. Это не нужно

VikkiRy replied:
Изменено. Настройки осуществляются данным VC самостоятельно

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -157,8 +188,57 @@ final class SignUpViewController: BaseViewController, SignUpViewProtocol {
+    }
+    
+    func updateCollectionView(toPage page: Int) {
+        let offset = pageControl.currentPage * Int(view.bounds.width)
```
воспользуйся scrollToIndexPath

VikkiRy replied:
Изменено

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -157,8 +188,57 @@ final class SignUpViewController: BaseViewController, SignUpViewProtocol {
+            return UICollectionViewCell()
+        }
+        
+        cell.imageView.image = presenter.images[indexPath.row]
```
скрой массив - предоставь функции для обращения за информацией

VikkiRy replied:
Изменено

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -8,17 +8,21 @@
     private lazy var appNameLabel = UILabel()
-    private lazy var logoImageView = UIImageView()
-    private lazy var logoLabel = UILabel()
+    private lazy var slidesCollectoinView = UICollectionView(frame: .zero, collectionViewLayout: UICollectionViewFlowLayout())
```
вызвав пустой конструктор вью проинициализируется теми же переменными

VikkiRy replied:
При такой инициализации возникает ошибка во время компиляции

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -74,22 +81,25 @@ final class SignUpViewController: BaseViewController, SignUpViewProtocol {
-        logoLabel.textAlignment = .left
-        logoLabel.textColor = Asset.Colors.logoLabelText.color
+        if let layout = slidesCollectoinView.collectionViewLayout as? UICollectionViewFlowLayout {
+            layout.scrollDirection = .horizontal
```
кажется можно конфигурируровать в UICollectionView,

VikkiRy replied:
Изменяется только в layout.scrollDirection

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -74,22 +81,25 @@ final class SignUpViewController: BaseViewController, SignUpViewProtocol {
+            let slidesEdgesInset = view.bounds.width / 2 - LayoutConstants.ImageLogo.widht / 2
+            
+            layout.minimumLineSpacing = slidesEdgesInset * 2
+            layout.sectionInset = UIEdgeInsets(top: 0, left: slidesEdgesInset, bottom: 0, right: slidesEdgesInset)
```
все переменные настраиваются в UICollectionView

VikkiRy replied:
Изменено

VikkiRy committed:
- BackgroundGradientView was refactored
- GradientView was removed
- LayoutConstants were updated
- Multipliers and Radius enums were created
- SignUpScreenPresenterProtocol, SignUpScreenPresenter, SignUpViewController were refactored

VikkiRy committed:
- HorizontalCollectionView was added
- CollectionView layout constants was updated
- slidesCollectoinView was updated

HKurylchykava commented on BadHabbit/BadHabbit/Kit/DesignKit/Models/Multipliers.swift:
```diff
@@ -0,0 +1,12 @@
+import Foundation
+
+enum Multipliers {
+    static let phraseLabelLineHeight = 1.5
```
why 1.5? we have the same value 1.26 on several screens

VikkiRy replied:
value was changed

VikkiRy committed:
HorizontalCollectionView class was removed

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/GradientButton.swift:
```diff
@@ -0,0 +1,86 @@
+            
+            if isEnabled {
+                layer.sublayers?.removeAll(where: { layer in
+                    layer == darkShadowLayer || layer == lightShadowLayer
```
Почему бы не вызвать remove from super у этих layer?

VikkiRy replied:
Проверка на наличие данных sublayers оставлена, принцип удаления изменен

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/GradientButton.swift:
```diff
@@ -0,0 +1,86 @@
+                layer.startPoint = .zero
+                layer.endPoint = .zero
+                layer.colors = []
+                layer.backgroundColor = Assets.Colors.startButtonUnenabledBackground.color.cgColor
```
Я бы вместо изменения удалил gradient layer

VikkiRy replied:
Готово

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/GradientButton.swift:
```diff
@@ -0,0 +1,86 @@
+
+import UIKit
+
+class GradientButton: UIButton {
```
Название не подходит, это уже не только про градиент

VikkiRy replied:
Изменено на ActionButton

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/Models/LayoutConstants.swift:
```diff
@@ -14,8 +14,24 @@ enum LayoutConstants {
         static let size: CGFloat = 196
     }
+    
+    enum CollectionView {
```
почему ты решила вынести в глобальные константы ?

VikkiRy replied:
Удалено

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/Models/Radius.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+enum Radius {
```
слишком общее название. Я бы назвал AppearanceConstants

VikkiRy replied:
Изменено имя и добавлен `enum CornerRadius`

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -74,44 +77,65 @@ final class SignUpViewController: BaseViewController, SignUpViewProtocol {
+        slidesCollectoinView.backgroundColor = .none
+        slidesCollectoinView.isPagingEnabled = true
+        slidesCollectoinView.showsHorizontalScrollIndicator = false
+        if let layout = slidesCollectoinView.collectionViewLayout as? UICollectionViewFlowLayout {
```
давай проинициализируем layout в при создании ( private lazy var layout: FlowLayout = {...}() )

VikkiRy replied:
Готово

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -153,12 +180,75 @@ final class SignUpViewController: BaseViewController, SignUpViewProtocol {
+    }
+    
+    private func updateStartButton(isEnabled: Bool) {
+        if isEnabled {
```
а зачем эта проверка?

VikkiRy replied:
Проверка в целом нужна так как у нас кнопка после отображения слайда должна оставаться активной по требованиям, но логика проверки была не верная. Изменено

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -153,12 +180,75 @@ final class SignUpViewController: BaseViewController, SignUpViewProtocol {
+            return UICollectionViewCell()
+        }
+        
+        cell.imageView.image = presenter.image(forRow: indexPath.row)
```
В ячейку надо передать данные - она сама будет конфигурировать

VikkiRy replied:
Если мы свойство `imageView` сделаем приватным, то нам нужно просто сделать инициализатор SignUpScreenCollectionViewCell, в параметрах которого будет UIImage? Немного не поняла как нужно сделать по-другому... 😞

nikGayko replied:
Добавить метод а-ля configure в Ячейку который будет принимать необходимые параметры для конфигурации этой ячейки. В этом методе останется только вызвать этот метод у ячейки

VikkiRy replied:
Спасибо, добавила

VikkiRy committed:
- GradientButton was renamed to ActionButton and refactored
- enum Radius was renamed to AppearanceConstants
- enum CollectionView was removed
- phraseLabelLineHeight value was changed
- setupViewConstraints func was removed
- SignUpViewController was refactored

VikkiRy committed:
configure func was added

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/ActionButton.swift:
```diff
@@ -0,0 +1,94 @@
+            Assets.Colors.darkBlue.color.cgColor,
+            Assets.Colors.purple.color.cgColor
+        ]
+        gradientLayer.frame = self.bounds
```
In case when bounds of button changes, what will be with bounds of gradient layer?

VikkiRy replied:
In my implementation - they wouldn't have changed.
Setting this property was moved to the draw method.

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/ActionButton.swift:
```diff
@@ -0,0 +1,94 @@
+                    if sublayer == darkShadowLayer || sublayer == lightShadowLayer {
+                        sublayer.removeFromSuperlayer()
+                    }
+                })
```
чем этот вызов отличается от
`darkShadowLayer.removeFromSuperlayer(); lightShadowLayer.removeFromSuperlayer()` ?

VikkiRy replied:
я какой-то момент, когда я меняла логику без этой проверки была некорректная работа. Но в текущей реализации - разницы нет. Изменила + в didSet поменяла конструкцию на switch.

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/ActionButton.swift:
```diff
@@ -0,0 +1,94 @@
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        
+        layer.masksToBounds = true
```
зачем это вызов?

VikkiRy replied:
Поскольку dark and light shadow path были определены как прямоугольник со смещением - нам нужно установить это св-во как true чтобы обрезались sublayers по biunds.
Либо менять shadow path и полностью отрисовывать путь...

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/Models/AppearanceConstants.swift:
```diff
@@ -0,0 +1,14 @@
+import Foundation
+
+enum AppearanceConstants {
+    enum CornerRadius {
```
я бы группировал не по свойству а по элементу

VikkiRy replied:
Изменено

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/Models/LayoutConstants.swift:
```diff
@@ -14,7 +14,9 @@ enum LayoutConstants {
     enum ActionButton {
         static let height = 50
         static let bottomInset = 20
+        static let topOffset = 30
```
почему topOffset в глобальных константах?

VikkiRy replied:
Удалено

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -11,16 +11,22 @@ import SnapKit
+    }()
     
-    private let presenter: SignUpScreenPresenterProtocol
+    lazy var startButton = ActionButton()
```
private

VikkiRy replied:
Изменено

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -153,7 +182,25 @@ final class SignUpViewController: BaseViewController, SignUpViewProtocol {
+    }
+    
+    private func updateStartButton() {
+        if !startButton.isEnabled {
```
по логике если пользователь пролистал всю карусель кнопка продолжить становиться доступной? И даже если он листает назад - мы ее не дизэйблим, так?

VikkiRy replied:
Да. Даже если пользователь возвращается на со следующего экрана на текущий - по требованиям отображается последний слайд и кнопка продолжает быть активной

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -153,7 +182,25 @@ final class SignUpViewController: BaseViewController, SignUpViewProtocol {
+            at: .centeredHorizontally,
+            animated: true)
+        
+        if sender.currentPage == 2 {
```
этот чек надо перенести в метод updateStartButton, он напрямую связан с состоянием кнопки

VikkiRy replied:
Перенесено

nikGayko commented on BadHabbit/BadHabbit/UserStories/SingUpScreen/SignUpViewController.swift:
```diff
@@ -162,3 +209,55 @@ final class SignUpViewController: BaseViewController, SignUpViewProtocol {
+        
+        pageControl.currentPage = indexPath.row
+        
+        if indexPath.row == 2 {
```
опять этот чек, значит точно надо перенести в функцию updateStartButton

nikGayko replied:
нельзя писать `2`, эта вью не знает размер карусели, размер карусели знает presenter

VikkiRy replied:
Перенесла и изменила

VikkiRy committed:
- ActionButton class was refactored
- CornerRadius enum renamed to ActionButton
- ActionButton topOffset LayoutConstant was removed
- updateStartButton() was refactored

HKurylchykava commented on BadHabbit/BadHabbit/Kit/DesignKit/ActionButton.swift:
```diff
@@ -0,0 +1,92 @@
+
+import UIKit
+
+class ActionButton: UIButton {
```
final

VikkiRy replied:
Done

VikkiRy committed:
ActionButton, lazy var imageView, let signInAppleScreenAssembly, let globalCoordinator were marked as private

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/ActionButton.swift:
```diff
@@ -0,0 +1,92 @@
+            case false:
+                if let sublayers = layer.sublayers,
+                   sublayers.contains(where: { $0 == gradientLayer }) {
+                    gradientLayer.removeFromSuperlayer()
```
а зачем перед removeFromSuperlayer проверяешь добавлен ли он как sublayer ?

VikkiRy replied:
Удалила проверку

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/Models/Multipliers.swift:
```diff
@@ -0,0 +1,12 @@
+import Foundation
+
+enum Multipliers {
+    static let phraseLabelLineHeight = 1.26
```
Добавь в AppearanceConstants

VikkiRy replied:
Добавлено

nikGayko commented on BadHabbit/BadHabbit/Kit/DesignKit/ActionButton.swift:
```diff
@@ -0,0 +1,92 @@
+    
+    override var isEnabled: Bool {
+        didSet {
+            switch isEnabled {
```
обычно все таки сравнивают булевы через if else, и впредь давай так и делать, это можешь оставить

VikkiRy replied:
Ок, не проблема, вернула if-else + установила `isEnabled = true` при инициализации (по принципу работы UIButton)

VikkiRy committed:
- ActionButton class was refactored
- enum Multipliers was removed

VikkiRy committed:
Merge branch 'develop' into feature/HAB-19-slides

# Conflicts:
#	BadHabbit/BadHabbit.xcodeproj/project.pbxproj
#	BadHabbit/BadHabbit/Kit/DesignKit/BackgroundGradientView.swift
#	BadHabbit/BadHabbit/Kit/DesignKit/GlobalCoordinator/GlobalCoordinator.swift
#	BadHabbit/BadHabbit/SupportFiles/Assets.xcassets/Colors/purple.colorset/Contents.json
#	BadHabbit/BadHabbit/SupportFiles/Assets.xcassets/Colors/purpleLight.colorset/Contents.json
#	BadHabbit/BadHabbit/SupportFiles/Assets.xcassets/Images/mountains.imageset/mountains.pdf
#	BadHabbit/BadHabbit/UserStories/SingnInAppleScreen/SignInAppleViewController.swift
